plugins/conversation-insights/src/index.ts

```diff
@@ -30,53 +34,52 @@ plugin.on.afterOutgoingMessage('*', async (props) => {
     return
   }
   const { conversation } = await props.client.getConversation({ id: props.data.message.conversationId })
-  await _onNewMessage({ ...props, conversation })
+  await onNewMessageHandler.onNewMessage({ ...props, conversation })
   return undefined
 })
 
-type OnNewMessageProps = CommonProps & {
-  conversation: bp.ClientOutputs['getConversation']['conversation']
-}
-const _onNewMessage = async (
-  props: OnNewMessageProps
-): Promise<{ message_count: number; participant_count: number }> => {
-  const message_count = props.conversation.tags.message_count ? parseInt(props.conversation.tags.message_count) + 1 : 1
-
-  const participant_count = await props.client
-    .listParticipants({ id: props.conversation.id })
-    .then(({ participants }) => participants.length)
-
-  const tags = {
-    message_count: message_count.toString(),
-    participant_count: participant_count.toString(),
-  }
-
-  await props.client.updateConversation({
-    id: props.conversation.id,
-    tags,
-  })
-  return { message_count, participant_count }
-}
-
 plugin.on.event('updateAiInsight', async (props) => {
   if (isBrowser) {
     props.logger.error('This event is not supported by the browser')
     return
   }
-  const firstMessagePage = await props.client
-    .listMessages({ conversationId: props.event.conversationId })
-    .then((res) => res.messages)
 
-  if (!props.event.conversationId) {
-    throw new sdk.RuntimeError(`The conversationId cannot be null when calling the event '${props.event.type}'`)
+  const workflows = await props.client.listWorkflows({
+    name: 'conversation-insights#updateAllConversations',
+    statuses: ['in_progress', 'listening', 'pending'],
+  })
+
+  if (workflows.workflows.length === 0) {
+    props.workflows.updateAllConversations.startNewInstance({ input: {} })
   }
-  const conversation = await props.client.getConversation({ id: props.event.conversationId })
+})
 
-  await summaryUpdater.updateTitleAndSummary({
-    ...props,
-    conversation: conversation.conversation,
-    messages: firstMessagePage,
-  })
+plugin.on.workflowStart('updateAllConversations', async (props) => {
+  props.logger.info('Starting updateAllConversations workflow')
+
+  return undefined
+})
+
+plugin.on.workflowContinue('updateAllConversations', async (props) => {
+  const dirtyConversations = await props.client.listConversations({ tags: { isDirty: 'true' } })
+
+  const promises = []
+  for (const conversation of dirtyConversations.conversations) {
+    const firstMessagePage = await props.client
+      .listMessages({ conversationId: props.event.conversationId })
+      .then((res) => res.messages)
+    promises.push(summaryUpdater.updateTitleAndSummary({ ...props, conversation, messages: firstMessagePage }))
+  }
+
+  await Promise.all(promises)
```

Comment:
~~**question:** What if we can't complete a single batch before the workflow times out? Maybe we should declare a workflow state and persist where we're at~~
**edit:** ok I think you kind of already persist that information with the `isDirty` tag